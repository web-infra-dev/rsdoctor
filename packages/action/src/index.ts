import { setFailed, getInput } from '@actions/core';
import { uploadArtifact } from './upload';
import { downloadArtifact, downloadArtifactByCommitHash } from './download';
import { GitHubService } from './github';
import {
  loadSizeData,
  generateSizeReport,
  getDemoBaselineData,
} from './report';
import path from 'path';

// Helper function to determine if this is a merge event
function isMergeEvent(): boolean {
  const { context } = require('@actions/github');
  return (
    context.eventName === 'push' &&
    context.payload.ref ===
      `refs/heads/${context.payload.repository.default_branch}`
  );
}

// Helper function to determine if this is a PR event
function isPullRequestEvent(): boolean {
  const { context } = require('@actions/github');
  return context.eventName === 'pull_request';
}

(async () => {
  try {
    const githubService = new GitHubService();

    const filePath = getInput('file_path');
    if (!filePath) {
      throw new Error('file_path is required');
    }
    const fullPath = path.resolve(process.cwd(), filePath);
    console.log(`Full path: ${fullPath}`);

    const fileName = path.basename(filePath);
    const relativePath = path.relative(process.cwd(), fullPath);
    const pathParts = relativePath.split(path.sep);
    const fileNameWithoutExt = path.parse(fileName).name;
    const fileExt = path.parse(fileName).ext;

    // Get current commit hash
    const currentCommitHash = githubService.getCurrentCommitHash();
    console.log(`Current commit hash: ${currentCommitHash}`);

    // Create artifact name pattern
    const artifactNamePattern = `${pathParts.join('-')}-${fileNameWithoutExt}-`;
    console.log(`Artifact name pattern: ${artifactNamePattern}`);

    if (isMergeEvent()) {
      // MR 合入时：只上传当前分支的工件
      console.log(
        '🔄 Detected merge event - uploading current branch artifact only',
      );

      const uploadResponse = await uploadArtifact(currentCommitHash, fullPath);

      if (typeof uploadResponse.id !== 'number') {
        throw new Error('Artifact upload failed: No artifact ID returned.');
      }

      console.log(
        `✅ Successfully uploaded artifact with ID: ${uploadResponse.id}`,
      );

      // Generate simple report for uploaded data
      const currentSizeData = loadSizeData(fullPath);
      if (currentSizeData) {
        await generateSizeReport(currentSizeData);
      }
    } else if (isPullRequestEvent()) {
      // MR 提交时：只下载目标分支的工件（如果存在）并生成比较报告
      console.log(
        '📥 Detected pull request event - downloading target branch artifact if exists',
      );

      // Load current size data
      const currentSizeData = loadSizeData(fullPath);
      if (!currentSizeData) {
        throw new Error(`Failed to load current size data from: ${fullPath}`);
      }

      let baselineSizeData: any = null;

      try {
        // Get target branch latest commit hash
        const targetCommitHash =
          await githubService.getTargetBranchLatestCommit();
        console.log(`Target branch commit hash: ${targetCommitHash}`);

        // Create artifact name for target branch (same naming pattern)
        const targetArtifactName = `${pathParts.join('-')}-${fileNameWithoutExt}-${targetCommitHash}${fileExt}`;
        console.log(`Looking for target artifact: ${targetArtifactName}`);

        // Try to find and download target branch artifact
        try {
          const downloadResult = await downloadArtifactByCommitHash(
            targetCommitHash,
            fileName,
          );
          const downloadedBaselinePath = path.join(
            downloadResult.downloadPath,
            fileName,
          );
          baselineSizeData = loadSizeData(downloadedBaselinePath);
          console.log('✅ Successfully downloaded target branch artifact');
        } catch (downloadError) {
          console.log(
            'ℹ️  No baseline data found - target branch artifact does not exist',
          );
          console.log('📝 Using demo baseline data for comparison');

          // Use built-in demo data as baseline for comparison
          baselineSizeData = getDemoBaselineData();
          console.log('✅ Successfully loaded demo baseline data');
        }
      } catch (error) {
        console.warn(`⚠️  Failed to get target branch commit: ${error}`);
        console.log('📝 Using demo baseline data for comparison');

        // Use built-in demo data as baseline for comparison
        baselineSizeData = getDemoBaselineData();
        console.log('✅ Successfully loaded demo baseline data');
      }

      // Generate report card
      await generateSizeReport(currentSizeData, baselineSizeData || undefined);
    } else {
      // 其他情况：默认行为（上传并尝试下载）
      console.log('🔄 Default behavior - uploading and downloading artifacts');

      const uploadResponse = await uploadArtifact(currentCommitHash, fullPath);

      if (typeof uploadResponse.id !== 'number') {
        throw new Error('Artifact upload failed: No artifact ID returned.');
      }

      console.log(
        `✅ Successfully uploaded artifact with ID: ${uploadResponse.id}`,
      );

      // Try to download target branch artifact
      try {
        const targetCommitHash =
          await githubService.getTargetBranchLatestCommit();
        console.log(`Target branch commit hash: ${targetCommitHash}`);

        const targetArtifactName = `${pathParts.join('-')}-${fileNameWithoutExt}-${targetCommitHash}${fileExt}`;
        console.log(`Looking for target artifact: ${targetArtifactName}`);

        const downloadResult = await downloadArtifactByCommitHash(
          targetCommitHash,
          fileName,
        );
        const downloadedBaselinePath = path.join(
          downloadResult.downloadPath,
          fileName,
        );
        const baselineSizeData = loadSizeData(downloadedBaselinePath);

        // Generate report card
        const currentSizeData = loadSizeData(fullPath);
        if (currentSizeData) {
          await generateSizeReport(
            currentSizeData,
            baselineSizeData || undefined,
          );
        }

        console.log('✅ Successfully downloaded target branch artifact');
      } catch (error) {
        console.warn(`⚠️  Failed to download target branch artifact: ${error}`);
        console.log('📝 Using demo baseline data for comparison');

        // Generate report card with demo baseline
        const currentSizeData = loadSizeData(fullPath);
        if (currentSizeData) {
          const demoBaseline = getDemoBaselineData();
          await generateSizeReport(currentSizeData, demoBaseline);
        }
      }
    }
  } catch (error) {
    if (error instanceof Error) {
      setFailed(error.message);
    } else {
      setFailed(String(error));
    }
  }
})();
