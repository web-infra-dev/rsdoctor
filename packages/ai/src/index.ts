export * from './mcp/index.js';
