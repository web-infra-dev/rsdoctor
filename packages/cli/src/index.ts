import yargs from 'yargs/yargs';
import { hideBin } from 'yargs/helpers';
import chalk from 'chalk';
import { Common } from '@rsdoctor/types';
import { analyze, bundleDiff } from './commands';
import { Command, CommandContext, GetCommandArgumentsType } from './types';
import { Commands, pkg, bin } from './constants';

export async function execute<
  T extends GetCommandArgumentsType<typeof analyze>,
>(
  command: Commands.Analyze | `${Commands.Analyze}`,
  options: T['options'],
): Promise<T['result']>;

export async function execute<
  T extends GetCommandArgumentsType<typeof bundleDiff>,
>(
  command: Commands.BundleDiff | `${Commands.BundleDiff}`,
  options: T['options'],
): Promise<T['result']>;

export async function execute(): Promise<void>;
export async function execute(
  command?: `${Commands}` | Commands,
  options?: Common.PlainObject,
): Promise<unknown> {
  const cwd = process.cwd();
  const { name, version } = pkg;

  const ctx: CommandContext = { bin, cwd, name };

  if (command === Commands.Analyze) {
    const { action } = analyze(ctx);

    return action(
      options as GetCommandArgumentsType<typeof analyze>['options'],
    );
  }

  if (command === Commands.BundleDiff) {
    const { action } = bundleDiff(ctx);

    return action(
      options as GetCommandArgumentsType<typeof bundleDiff>['options'],
    );
  }

  const argv = hideBin(process.argv);
  const args = yargs(argv).usage(`${bin} <command> [options]`);

  args.version(version);

  const commands: Command<string>[] = [analyze, bundleDiff];

  commands.forEach((cmd) => {
    const { command, description, options, action } = cmd(ctx);

    args.command(
      command,
      description,
      (yargs) => {
        return options(yargs.usage(`${bin} ${command} [options]`));
      },
      async (args) => {
        try {
          await action(args);
        } catch (error) {
          const { message, stack } = error as Error;
          console.log('');
          console.error(chalk.red(stack || message));
          process.exit(1);
        }
      },
    );
  });

  if (!argv.length) {
    args.showHelp();
  }

  await args.parse();
}
