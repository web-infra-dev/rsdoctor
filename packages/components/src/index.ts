export * from './components';
export * from './utils';
export * as Constants from './constants';