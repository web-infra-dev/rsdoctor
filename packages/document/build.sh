#!/bin/bash
set -e

echo "🚀 Starting build script..."

dir=`dirname $0`
rootdir=$(cd $dir && pwd)
outdir="$rootdir"/doc_build

echo "📁 Working directory: $rootdir"
echo "📁 Output directory: $outdir"

# copy client/dist
echo "📦 Copying client/dist to preview folder..."
mkdir -p "${outdir}/preview"
client_dist_path="$(cd $dir && cd ../../ && pwd)/packages/client/dist"
echo "📂 Source path: $client_dist_path"

if [ -d "$client_dist_path" ]; then
  echo "✅ Client dist directory exists"
  cp -rf "$client_dist_path/"* "${outdir}/preview/"
  echo "✅ Successfully copied client files to preview folder"
  
  # Show what was copied
  echo "📋 Preview folder contents:"
  ls -la "${outdir}/preview/"

  echo "🎉 Build script completed successfully!"
else
  echo "❌ Error: Client dist directory not found at $client_dist_path"
fi



