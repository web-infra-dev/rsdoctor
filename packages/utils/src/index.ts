export * as RuleUtils from './rule-utils';
export * as Logger from './logger';
