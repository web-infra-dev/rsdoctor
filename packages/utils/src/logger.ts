import chalk from 'chalk';
import { logger } from 'rslog';
import { Constants } from '@rsdoctor/types';

/**
 * log debug message
 */
export function debug(getMsg: () => string, prefix = '') {
  if (!process.env[Constants.RsdoctorProcessEnvDebugKey]) {
    return;
  }

  logger.level = 'verbose';
  logger.debug(`${prefix} ${getMsg()}`);
}

export { chalk, logger };
