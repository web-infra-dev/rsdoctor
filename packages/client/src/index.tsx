import '@rsdoctor/components/i18n';
import App from './App';
import ReactDOM from 'react-dom/client';
import './common/styles/base.scss';

const rootElement = document.getElementById('root');
const root = ReactDOM.createRoot(rootElement!);

root.render(<App />);

const link = document.createElement('link');
link.setAttribute('type', 'image/x-icon');
link.setAttribute('rel', 'icon');
link.setAttribute(
  'href',
  'https://assets.rspack.dev/rsdoctor/rsdoctor-logo-960x960.png',
);
document.head.appendChild(link);
