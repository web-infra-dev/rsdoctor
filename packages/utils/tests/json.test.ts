import { describe, it, expect } from 'vitest';
import { Json } from '../src/build';

describe('test src/json.ts', () => {
  describe('stringify()', () => {
    it('string & number & null & undefined & boolean', async () => {
      expect(await Json.stringify('abcde')).toEqual('"abcde"');
      expect(await Json.stringify(123)).toEqual('123');
      expect(await Json.stringify(null)).toEqual('null');
      expect(await Json.stringify(undefined)).toEqual(undefined);
      expect(await Json.stringify(true)).toEqual('true');
      expect(await Json.stringify(false)).toEqual('false');
    });

    it('Array & Object', async () => {
      expect(await Json.stringify(['abcde'])).toMatchSnapshot();
      expect(
        await Json.stringify(['abcde', 123, null, undefined, true, false]),
      ).toMatchSnapshot();
      expect(
        await Json.stringify([
          { a: 1, b: undefined, c: null },
          1,
          [2, { k: 1 }],
        ]),
      ).toMatchSnapshot();

      expect(
        await Json.stringify({ a: 1, b: undefined, c: null }),
      ).toMatchSnapshot();
      expect(
        await Json.stringify({ a: 1, b: undefined, c: null, d: { e: 23 } }),
      ).toMatchSnapshot();
      expect(
        await Json.stringify({
          d: { e: 23, f: null, g: undefined, h: { a: 1 } },
        }),
      ).toMatchSnapshot();
    });
  });
});
