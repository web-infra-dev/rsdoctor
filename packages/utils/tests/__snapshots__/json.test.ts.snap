// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test src/json.ts > stringify() > Array & Object 1`] = `
[
  "["abcde"]",
]
`;

exports[`test src/json.ts > stringify() > Array & Object 2`] = `
[
  "["abcde",123,null,null,true,false]",
]
`;

exports[`test src/json.ts > stringify() > Array & Object 3`] = `
[
  "[{"a":1,"c":null},1,[2,{"k":1}]]",
]
`;

exports[`test src/json.ts > stringify() > Array & Object 4`] = `
[
  "{"a":1,"c":null}",
]
`;

exports[`test src/json.ts > stringify() > Array & Object 5`] = `
[
  "{"a":1,"c":null,"d":{"e":23}}",
]
`;

exports[`test src/json.ts > stringify() > Array & Object 6`] = `
[
  "{"d":{"e":23,"f":null,"h":{"a":1}}}",
]
`;
